pkg/controller/mysql/grant/reconciler.go

```diff
@@ -0,0 +1,324 @@
+		return nil, errors.Wrap(err, errGetPC)
+	}
+
+	// We don't need to check the credentials source because we currently only
+	// support one source (MySQLConnectionSecret), which is required and
+	// enforced by the ProviderConfig schema.
+	ref := pc.Spec.Credentials.ConnectionSecretRef
+	if ref == nil {
+		return nil, errors.New(errNoSecretRef)
+	}
+
+	s := &corev1.Secret{}
+	if err := c.kube.Get(ctx, types.NamespacedName{Namespace: ref.Namespace, Name: ref.Name}, s); err != nil {
+		return nil, errors.Wrap(err, errGetSecret)
+	}
+
+	return &external{
+		db:   c.newDB(s.Data),
+		kube: c.kube,
+	}, nil
+}
+
+type external struct {
+	db   xsql.DB
+	kube client.Client
+}
+
+func (c *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.Grant)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotGrant)
+	}
+
+	username := *cr.Spec.ForProvider.User
+	dbname := *cr.Spec.ForProvider.Database
+
+	privileges, result, err := c.getPrivileges(ctx, username, dbname)
+	if err != nil {
+		return managed.ExternalObservation{}, err
+	}
+	if result != nil {
+		return *result, nil
+	}
+
+	if !comparePrivileges(cr.Spec.ForProvider.Privileges, privileges) {
+		return managed.ExternalObservation{
+			ResourceExists:   true,
+			ResourceUpToDate: false,
+		}, nil
+	}
+
+	cr.SetConditions(runtimev1alpha1.Available())
+
+	return managed.ExternalObservation{
+		ResourceExists:   true,
+		ResourceUpToDate: true,
+	}, nil
+}
+
+func comparePrivileges(a, b []string) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	sort.Strings(a)
+	sort.Strings(b)
+
+	for i := range a {
+		// Special case because ALL is an alias for "ALL PRIVILEGES"
+		strA := strings.ReplaceAll(a[i], allPrivileges, "ALL")
+		strB := strings.ReplaceAll(b[i], allPrivileges, "ALL")
+		if strA != strB {
+			return false
+		}
+	}
+	return true
+}
+
+func parseGrant(grant, dbname string) (privileges []string) {
+	matches := grantRegex.FindStringSubmatch(grant)
+	if len(matches) == 3 && matches[2] == dbname {
+		return strings.Split(matches[1], ", ")
+	}
+	return nil
+}
+
+func (c *external) getPrivileges(ctx context.Context, username, dbname string) ([]string, *managed.ExternalObservation, error) {
+	username, host := mysql.SplitUserHost(username)
+	query := fmt.Sprintf("SHOW GRANTS FOR %s@%s", mysql.QuoteValue(username), mysql.QuoteValue(host))
+	rows, err := c.db.Query(ctx, xsql.Query{String: query})
+	if err != nil {
+		return nil, nil, errors.Wrap(err, errCurrentGrant)
+	}
+	defer rows.Close() //nolint:errcheck
+
+	var privileges []string
+	for rows.Next() {
+		var grant string
+		if err := rows.Scan(&grant); err != nil {
+			return nil, nil, errors.Wrap(err, errCurrentGrant)
+		}
+		p := parseGrant(grant, dbname)
+		if p != nil && privileges != nil {
+			// Found more than one grant for this user on this DB
```

Comment:
Do we support creating multiple Grant resources for one User and Database? Should we?

Reply:
Currently only a single `Grant` resource can exist for every user/database combination. Multiple privileges on a single Grant are possible though.
This place in the code basically assumes some external system changed the grants and we should update them again according to the controller's state of truth.

I don't see a way to support multiple Grant resources for the same user/database combination since we couldn't differentiate which MySQL grant is coming from which `Grant` resource. Also as far as I saw, MySQL combines the privileges of multiple `GRANT` queries for the same user/database combo.

Reply:
Got it. It might be a good idea to explain this in a comment - it could be obvious to those more versed in SQL but it was not to me.